lessons/module-2/introduction-to-typescript.md

```diff
@@ -0,0 +1,320 @@
+4. **Scalability**: TypeScript is particularly useful for large codebases, where it can help manage complexity.
+
+## Comparing JavaScript and TypeScript
+JavaScript is dynamically typed, meaning variable types are determined at runtime. TypeScript, on the other hand, allows for static type checking, where types are checked at compile-time.
+
+<section class="call-to-action">
+### Example of JavaScript vs TypeScript
+
+**JavaScript:**
+```js
+let greeting = "Hello, world!";
+greeting = 42; // No error in JavaScript, but might cause issues later
+console.log(greeting); // Outputs: 42
+```
+
+**TypeScript:**
+```ts
+let greeting: string = "Hello, world!";
+greeting = 42; // Error in TypeScript
+console.log(greeting); // Outputs: 42
+```
+
+1. What potential issues might arise from the JavaScript example where `greeting` is reassigned to a number? How does TypeScript help prevent these issues?
+2. Consider a large codebase with multiple developers. How might TypeScript's type system improve collaboration and code maintenance compared to JavaScript?
+</section>
+
+## Type Annotations
+TypeScript allows you to annotate variables with types including all of the standard data types you have used in JavaScript like strings, numbers, booleans, arrays, and objects. This helps catch type errors at compile time.
+
+Here are some examples of type annotations in TypeScript:
+
+```ts
+let age: number = 25;
+let name: string = "Alice";
+let isStudent: boolean = true;
+let hobbies: string[] = ["reading", "hiking"];
+let person: { name: string; age: number } = { name: "Bob", age: 30 };
+```
+
+<section class="call-to-action">
+### Your Turn
+
+**Example 1**
+```js
+let title = "TypeScript Basics";
+let duration = 60;
+let isComplete = false;
+```
+
+* Convert the above JavaScript code to TypeScript by adding appropriate type annotations.
+
+**Example 2**
+```ts
+let score: number = "high";
+```
+* What error will the above TypeScript code produce?
+</section>
+
+<section class="answer">
+### Solution
+1. The solution to converting the variables would look like:
+```ts
+let title: string = "TypeScript Basics";
+let duration: number = 60;
+let isComplete: boolean = false;
+```
+2. The error will be a compile-time error because the type of the variable `score` is not a number.
+</section>
+
+<section class="note">
+### A Note about `any`
+
+TypeScript also supports a type called `any`, which means that a variable can be any type.  This is useful when you want to store a value of any type in a variable.  However, using `any` should be avoided because it disables type-safety.
+
+An example of using `any` would look like:
+```ts
+let id: any = 101;
+id = "E101"; // No error
+```
+</section>
+
+### Union Types
+In addition to type annotation, TypeScript also supports union types.  Union types allow a variable to hold more than one type.  Let's take a look at an example:
```

Comment:
I like this example too! Feels like a realistic case for having two possibel data types. And this is way better than `any`!